src/main/java/com/zendesk/maxwell/MaxwellUpdateRowsEvent.java

```diff
@@ -48,4 +52,64 @@ public String sqlOperationString() {
 	public String getType() {
 		return "update";
 	}
+
+	private LinkedList<Pair<Row>> filteredRowsBeforeAndAfter;
+	private boolean performedBeforeAndAfterFilter;
+
+	private List<Pair<Row>> filteredRowsBeforeAndAfter() {
+		if ( this.filter == null)
+			return event.getRows();
+
+		if ( performedBeforeAndAfterFilter )
+			return filteredRowsBeforeAndAfter;
+
+		filteredRowsBeforeAndAfter = new LinkedList<>();
+		for ( Pair<Row> p : event.getRows()) {
+			if ( this.filter.matchesRow(this, p.getAfter()) )
+				filteredRowsBeforeAndAfter.add(p);
+		}
+		performedBeforeAndAfterFilter = true;
+		return filteredRowsBeforeAndAfter;
+	}
+
+	@Override
+	public List<RowMap> jsonMaps() {
+		ArrayList<RowMap> list = new ArrayList<>();
+		Object avalue;
+		Object bvalue;
+		for (Pair<Row> p : filteredRowsBeforeAndAfter() ) {
+			Row after = p.getAfter();
+			Row before = p.getBefore();
+
+			RowMap rowMap = buildRowMap();
+
+			Iterator<Column> aftIter = after.getColumns().iterator();
+			Iterator<Column> befIter = before.getColumns().iterator();
+			Iterator<ColumnDef> defIter = table.getColumnList().iterator();
+			while ( aftIter.hasNext() && defIter.hasNext() && befIter.hasNext() ) {
+				Column afterColumn = aftIter.next();
+				ColumnDef columnDef = defIter.next();
+				Column beforeColumn = befIter.next();
+
+				avalue = valueForJson(afterColumn);
+
+				bvalue = valueForJson(beforeColumn);
+
+				if (avalue != null) {
+					avalue = columnDef.asJSON(avalue);
+					if ( !avalue.equals(columnDef.asJSON(bvalue)) ) {
```

Comment:
1.  this code still feels a bit hard to read. 
2.  I think `columnDef.asJSON(bvalue)` will probably crash if `bvalue` is null -- you need the null check before this.